rfc/accepted/0021-command-driven-connectors.md

```diff
@@ -0,0 +1,126 @@
+  end;
+  
+  create pipeline main
+  pipeline
+    select { "command": "read", "path": event.path } from in into out;
+  end;
+  
+  create connector read_file_list from read_file_list;
+  create connector file_connector from file_connector;
+  create connector file_list from file_list;
+  
+  connect /connector/read_file_list/out to /connector/file_list/control;
+  connect /connector/file_list/out to /pipeline/main/in;
+  
+  # This is the magic - we send the commands here, note the "control" port
+  connect /pipeline/main/out to /connector/file_connector/control;
+  
+  connect /connector/file_connector/out to /pipeline/main/out;
+end;
+```
+
+## Reference-level Explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Each command driven connector implements at least one port - `control`.
+`control` is an input port, through which the commands are sent.
+
+The commands are standardised as far as it is practical, so the connectors can be swapped without adjusting the rest of the system.
+Convenience functions are provided to generate the events.
+Each event has a `correlation` field, which will be a unique identifier provided in the metadata for the response. This is automatically generated by the convenience functions.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This raises the complexity of Tremor.
+
+## Rationale and Alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There are no known alternatives that provide the same benefits.
+Currently, for example the S3 reader connector, will read all files in the bucket, once, which has limited use.
+
+## Prior Art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and the bad, in relation to this proposal.
+A few examples of what this can include are:
+
+- For language, library, tools, and clustering proposals: Does this feature exist in other programming languages, and what experience have their community had?
+- For community proposals: Is this done by some other community and what were their experiences with it?
+- For other teams: What lessons can we learn from what other communities have done here?
+- Papers: Are there any published papers or great posts that discuss this? If you have some relevant papers to refer to, this can serve as a more detailed theoretical background.
+
+This section is intended to encourage you as an author to think about the lessons from other projects, provide readers of your RFC with a fuller picture.
+If there is no prior art, that is fine- your ideas are interesting to us whether they are brand new or if it is an adaptation from other projects.
+
+Note that while precedent set by other projects is some motivation, it does not, on its own, motivate an RFC.
+Please also take into consideration that Tremor sometimes intentionally diverges from similar projects.
+
+## Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- How do we enforce uniformity of commands across connectors?

```

Comment:
I really like the concept of `traits`, `behaviors`,  `personas` , or whatever it would be called for that. (I'll go with traid in the rest of the text as it's the shortest, which is pointless since I probably wrote more in this note then I save by using trait ...)

The basic idea is that a connector defines a number of traits that specify the kind of commands it interacts with. Which traits exist we got to define possibly not in the RFC as it'll change over time, but we could start with some examples.

lets look at s3 / file / gcs as an example. I suspect it would work something like this:


* **s3** offers the `fileio`, `objectstore` and `s3` traits
* **gcs** offers the `fileio`, `objectstore` and `gcs` traits
* **file** offers the `fileio`, `fs` traits

commands such as `list files`, `open file`, `close file` etc would be in `fileio`

commands like `cache id` could be in `objectstore`

and endpoint specific commands could be in `gcs` and `s3` respectively

The reason for this is that with generic traitrs testing / prototyping and migrating become very easy. As long as no implementation specific features are used prototyping could happen with a local only connector and then switched over to a production connector on deployment.

Reply:
Yes - reusable traits could be a huge boon for testing. For example, if we added `seek` or `sync` to `fileio` or other commands to `fileio` then these would be useable in tests for setup and may imply a set of associated assertions.